### PR TITLE
NOD Campaign maps (1 - 8)

### PR DESCRIPTION
**Pull request by kesh52**
This is an implementation of NOD Campaign maps.

Maps 1 and 3 fixed and updated. Maps 2, 4 - 8 created from scratch.

Some triggers were not implemented, because there is no native implementation in OpenRA project for them.
Link for trigger-implementation details: https://scm.thm.de/redmine/projects/dtdj-ultj/wiki
This WiKi page is on german. If you need a translation, please write an email to: alexander.boll@mni.thm.de

### CODE REVIEW

**Comment by penev92** [comment]
OK, to start off slowly:
I personally think this is *awesome*, because we have very few people who work on the missions.
Having said that, expect *a lot* of changes to be requested on this.
I suppose the first two would be to file a separate PR for each mission and to rebase properly on latest bleed.
Then come the indentation and style fixes, testing, feature/bug-fix changes.
Get ready for a long road ahead if you want this merged - after all, those are 16 missions.
I hope you have the patience to see this through ;)

**Comment by abcdefg30** [comment]
You need to undo the map format changes.

**Comment by obrakmann** [comment]
First of all, I think it's awesome that (1) your prof gave you the chance to work on this for your studies, and (2) that you managed to accomplish that much in the little time that was given to you.

However, you based your work on the last release, and a lot has changed since then. At the current stage this will need a lot more work to be merge-able.

Most importantly, this PR is way too big. Please submit one mission per PR, one PR at a time, preferably starting with nod02. Then you will need to incorporate all the map changes that have been introduced since the last release. In particular, there is a new binary map format, and the map.yaml files have received changes as well . The map.bin file will get updated automatically when the map is first run using current bleed (or the -playtest, even), the map.yaml file will need manual editing.

More required changes from a quick look:
* We try to adhere to our [coding style](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard) as much as possible (as applicable to the Lua language)
* You can hardcode the videos used for a mission in the map.yaml file now
* We generally disable creeps/viceroids in missions (using the `Creeps: False` map option)
* We disable the newly-introduced short game option on all missions (using the `ShortGame: False` map option). This makes it (almost) unnecessary to mess with the `MustBeDestroyed` trait.
* Unused player references (such as `Multi0` etc.) should be removed.
* The player stances (Enemy, Neutral, etc.) will need refinements. Civilians should always be an enemy to Nod.
* Named actors in the map.yaml files that are referenced in the lua script usually get a custom name. Those actors should go to the end of the actors list in map.yaml.
* Some missions need to prevent the refinery from spawning an additional harvester, as there's already a pre-placed one. Look at how other missions do it (keyword is `FreeActor`).
* The tech tree often doesn't seem to be correct (eg. the Temple of Nod is buildable in mission 5).
* Some missions (nod04*, for example) fail immediately at the start.
* Some missions stuff multiple objectives into a single one. These will need to be split.

As I mentioned, this will need a lot more work, and I would totally understand if you're just hoping to complete your given task and not bother with this. Bud of course we hope that you'd rather acquired a taste for working on these missions, and we would be very pleased if you continued, since we're seriously understaffed in the mapping department :).

Anyway, thank you very much for your contribution, and please let us know how you choose to proceed.

File: mods/cnc/maps/nod01/map.yaml [hunk 1]
```diff
@@ -1,38 +1,35 @@
-MapFormat: 7
+Selectable: False
+
+MapFormat: 6
```
Comment (Mailaender) [inline mods/cnc/maps/nod01/map.yaml]: This reverts https://github.com/OpenRA/OpenRA/pull/7134 which we can't accept.

File: mods/cnc/maps/nod01/map.yaml [hunk 2]
```diff
@@ -1,38 +1,35 @@
...
 
-Visibility: MissionSelector
+UseAsShellmap: False
 
 Type: Campaign
 
-Videos:
-	BackgroundInfo: intro2.vqa
-	Briefing: nod1.vqa
-	GameLost: nodlose.vqa
```
Comment (Mailaender) [inline mods/cnc/maps/nod01/map.yaml]: This needs to be undone. See https://github.com/OpenRA/OpenRA/pull/7072 for comparison.

File: mods/cnc/maps/nod01/map.yaml [hunk 3]
```diff
@@ -1,38 +1,35 @@
...
-
 Options:
 	Crates: False
-	Creeps: False
```
Comment (Mailaender) [inline mods/cnc/maps/nod01/map.yaml]: Please don't delete this. It was introduced in https://github.com/OpenRA/OpenRA/pull/7220.

File: mods/cnc/maps/nod01/map.yaml [hunk 4]
```diff
@@ -1,38 +1,35 @@
...
 	Fog: True
 	Shroud: True
 	AllyBuildRadius: False
 	FragileAlliances: False
 	StartingCash: 0
 	ConfigurableStartingUnits: False
-	ShortGame: False
```
Comment (Mailaender) [inline mods/cnc/maps/nod01/map.yaml]: This was introduced in https://github.com/openra/openra/pull/6964 so don't remove that.

**Comment by Mailaender** [comment]
This is impressive work. I can only imagine how many man-hours it took you. I linked the pull requests that you missed while working on this at your private fork so you get an understanding how the engine evolved over the last months. You definitely have to split this into several per-mission pull requests. This is too huge. https://github.com/OpenRA/OpenRA/pull/7387/files takes very long to load and it will be very much work at once to rebase this properly. Also it is impossible for us to test and review all 8 missions at once.

**Comment by 42foobar42** [comment]
Hey guys,
I am Alex and I also worked on this project. I did a fork with my github account and will do a PR when a fixed map number 2a according to your comments. I do not know if I/we have enough time to fix all the maps, but I will try it for this mission and we will see how it works out.

By the way there is another issue according to the missions. As far as I know it is not possible to give your opponent in AI in the missions. So the opponent acts really stupid in the missions, but you can configure an AI in the multiplayer maps. Will it be possible to configure an AI in the mission in the future, because I think it is very inconvenient to create an AI through lua.

**Comment by Mailaender** [comment]
That should be fairly simple. Just like the skirmish lobby you have to assign a bot controller to a player or trigger it manually. See https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Player.cs#L91

**Comment by penev92** [comment]
Keep in mind that adding uncontrolled AIs to missions might have unwanted side effects to the mission.

**Comment by obrakmann** [comment]
Hi Alex,

that's great to hear!  If you have any questions or need any help, feel free to join us in our irc channel (late afternoon and evening would be your best chance to catch one of the mappers).

Our AI is really stupid and doesn't allow the fine-grained control that mission scripts need. In fact, there are long-term plans to move the AI code to lua completely, see #4947.

I'm going to close this PR now and we'll wait for yours for nod02a. Thanks again!

**Comment by Mailaender** [comment]
I assume these scripts were auto-generated from triggers, but can't find the source code of the utility. https://scm.thm.de/redmine/projects/dtdj-ultj/wiki/Automatische_Generierung_von_Karten Was this just planned for the future or did you succeed with it?

**Comment by 42foobar42** [comment]
Some of this points were implemented. The student who did this also want to deliver this to the project. I will ask him if he is making an PR.

**Pull request closed without merging**